junit-platform-launcher/src/main/java/org/junit/platform/launcher/tagexpression/ShuntingYard.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2015-2017 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.launcher.tagexpression;
+
+import static java.lang.Integer.MIN_VALUE;
+import static org.junit.platform.launcher.tagexpression.Expressions.tag;
+import static org.junit.platform.launcher.tagexpression.Operator.nullaryOperator;
+import static org.junit.platform.launcher.tagexpression.ParseStatus.emptyTagExpression;
+import static org.junit.platform.launcher.tagexpression.ParseStatus.missingClosingParenthesis;
+import static org.junit.platform.launcher.tagexpression.ParseStatus.missingOpeningParenthesis;
+import static org.junit.platform.launcher.tagexpression.ParseStatus.missingOperator;
+import static org.junit.platform.launcher.tagexpression.ParseStatus.success;
+
+import java.util.List;
+
+/**
+ * This is based on a modified version of the <a href="https://en.wikipedia.org/wiki/Shunting-yard_algor">Shunting-yard algorithm</a>
+ */
+class ShuntingYard {
+	private static final Operator RightParenthesis = nullaryOperator(")", -1);
+	private static final Operator LeftParenthesis = nullaryOperator("(", -2);
+	private static final Operator Sentinel = nullaryOperator("sentinel", MIN_VALUE);
+
+	private final Operators validOperators = new Operators();
+	private final Stack<Position<Expression>> expressions = new DequeStack<>();
+	private final Stack<Position<Operator>> operators = new DequeStack<>();
+
+	private final List<String> tokens;
+
+	ShuntingYard(List<String> tokens) {
+		this.tokens = tokens;
+		pushOperatorAt(-1, Sentinel);
+	}
+
+	public ParseResult execute() {
+		// @formatter:off
+		ParseStatus parseStatus = processTokens()
+                .process(this::consumeRemainingOperators)
+                .process(this::ensureOnlySingleExpressionRemains);
+        // @formatter:on
```

Comment:
Indentation looks wrong here. Spaces instead of tabs?

Reply:
replaced with tabs